GSoC 2020: measure repeats

IsaacWeiss opened the pull request:
Resolves: *https://musescore.org/en/node/10220*

<!-- Use "x" to fill the checkboxes below like [x] -->

- [x] I signed [CLA](https://musescore.org/en/cla)
- [x] I made sure the code in the PR follows [the coding rules](https://github.com/musescore/Documentation/blob/master/WorkflowAndGuidelines/CodeGuidelines.md)
- [x] I made sure the code compiles on my machine
- [x] I made sure there are no unnecessary changes in the code
- [x] I made sure the title of the PR reflects the core meaning of the issue you are solving
- [x] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"
- [x] I made sure the commit message title starts with "fix #424242:" if there is a related issue
- [x] I created the test (mtest, vtest, script test) to verify the changes I made

IsaacWeiss commented on libmscore/layout.cpp:
```diff
@@ -3602,38 +3602,24 @@ System* Score::collectSystem(LayoutContext& lc)
             breakMeasure = lc.curMeasure;
-            if (lc.prevMeasure->noBreak() && system->measures().size() > 2) {
-                // TODO: why 2? should this really be the total size of the nobreak group?
-                // to make sure we don't remove everything?
```
@MarcSabatella This was a typo, should have been 1. But, as you also pointed out, the `if/else` wasn't even necessary.

IsaacWeiss commented on libmscore/layout.cpp:
```diff
@@ -3602,38 +3602,24 @@ System* Score::collectSystem(LayoutContext& lc)
-                    minWidth -= system->lastMeasure()->width();
-                    system->removeLastMeasure();
-                    // TODO: this seems kludgy, seems like it means we set for the same measure
-                    // on two separate iterations of the loop
```
@MarcSabatella Absolutely right.

IsaacWeiss commented on libmscore/layout.cpp:
```diff
@@ -3602,38 +3602,24 @@ System* Score::collectSystem(LayoutContext& lc)
-                break;
-            } else if (!lc.prevMeasure->noBreak()) {
-                // TODO: is it really necessary to have a separate clause here?
-                // it seems we should be able to handle this within the if clause above
```
@MarcSabatella Yep. :-)

IsaacWeiss commented on libmscore/layout.cpp:
```diff
@@ -3735,7 +3721,7 @@ System* Score::collectSystem(LayoutContext& lc)
                 }
-                while (true) {
+                for (;;) {
                     // TODO: what if the nobreak group takes the entire system - is this correct?
```
@MarcSabatella I think the answer is yes, but why might it not be?

MarcSabatella replied:
No idea, but somehow it seems like something to be sure to test.

MarcSabatella commented on mu4/domain/importexport/internal/musicxml/importmxmlpass2.cpp:
```diff
@@ -2114,6 +2117,33 @@ void MusicXMLParserPass2::measure(const QString& partId,
+    // measure repeat handling
+    int staffIdx = _score->staffIdx(part);
+    if (_measureRepeatsStarted) {
+        // TODO: delete anything already added to measure
```
There are a few functions that are used for this sort of thing, but to be honest I never remember which is which.  One might be makeGap().  But what I would do is trace the code that is executed when contents of a measure are cleared for other reason, like if you select the measure and hit Delete, or if you try to enter a whole note into a measure that was already full of stuff.  Probably the latter is more relevant to your case.  Set a breakpoint on Score::setNoteRest() and single-step away!

IsaacWeiss commented:
Makes sense, thanks.

On Mon, Aug 3, 2020, 12:49 PM Marc Sabatella <notifications@github.com>
wrote:

> *@MarcSabatella* commented on this pull request.
> ------------------------------
>
> In mu4/domain/importexport/internal/musicxml/importmxmlpass2.cpp
> <https://github.com/musescore/MuseScore/pull/6365#discussion_r464534788>:
>
> > @@ -2114,6 +2117,33 @@ void MusicXMLParserPass2::measure(const QString& partId,
>          measure->setBreakMultiMeasureRest(true);
>      }
>
> +    // measure repeat handling
> +    int staffIdx = _score->staffIdx(part);
> +    if (_measureRepeatsStarted) {
> +        // TODO: delete anything already added to measure
>
> There are a few functions that are used for this sort of thing, but to be
> honest I never remember which is which. One might be makeGap(). But what I
> would do is trace the code that is executed when contents of a measure are
> cleared for other reason, like if you select the measure and hit Delete, or
> if you try to enter a whole note into a measure that was already full of
> stuff. Probably the latter is more relevant to your case. Set a breakpoint
> on Score::setNoteRest() and single-step away!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/musescore/MuseScore/pull/6365#pullrequestreview-460182173>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACYT7GLG3DDLH3FB6POST6TR63TA7ANCNFSM4PFIB5RA>
> .
>

IsaacWeiss commented:
For the record, the answer to that "(?)" in the latest commit message is a no. Doesn't produce correct output if the MeasureRepeat is in a multi-staff instrument (@MarcSabatella, testing and insights on this welcome).

MarcSabatella commented on mu4/domain/importexport/internal/musicxml/exportxml.cpp:
```diff
@@ -5053,18 +5054,43 @@ static bool elementRighter(const Element* e1, const Element* e2)
+    } else {
+        Part* part = m->score()->parts().at(partIndex);
+        for (int i = 0; i < part->nstaves(); ++i) {
+            int staffIdx = part->staff(i)->idx();
```
Probably there is some slightly more clever way of dealing with this - if you look at the implementation of idx(), you'll see it's relying on indexOf, which is probably slow operation.  Not that it really matters, and I know I told you this was the way to do it.  But still, you asked for feedback, that's what I've got... :-)

IsaacWeiss replied:
Improved efficiency by only calling it once, before the loop commences.

MarcSabatella commented on mu4/domain/importexport/internal/musicxml/exportxml.cpp:
```diff
@@ -6655,20 +6701,18 @@ void ExportMusicXml::writeParts()
 
-                    // write the measure or, in case of a multi measure rest,
-                    // the measure range it replaces
+                    int staffIdx = partIndex; // this is WRONG, only a placeholder until I figure out how to get correct staffIdx
```
This still needs to be corrected?

IsaacWeiss replied:
Yep

MarcSabatella commented on mu4/domain/importexport/internal/musicxml/exportxml.cpp:
```diff
@@ -6572,7 +6618,7 @@ void ExportMusicXml::writeMeasure(const Measure* const m,
 
     // output attribute at start of measure: measure-style
-    measureStyle(_xml, _attr, m);
+    measureStyle(_xml, _attr, m, partIndex);
```
So, previously the same measure style got written for each staff?

IsaacWeiss replied:
Yes, that's right. It actually still is, because I haven't yet accounted for the "number" attribute that may or may not be present in the <measure-style> tag, but this is a WIP.

IsaacWeiss replied:
Done.

IsaacWeiss commented:
Some notes on MusicXML handling of measure repeats in Sibelius (Ultimate 2020.6), Finale (26.3.1.520), and MuseScore (this branch):

- Sibelius is just bad. It doesn't include the `"number"` attribute in the `<measure-style>` tag surrounding the `<measure-repeat>` tag (for when a part/instrument has more than one staff and the measure repeat is not the same in all staves); it doesn't write the `"slashes"` attribute; it can't even import its own exports.
- Finale is better, unsurprisingly; I think MusicXML started there. Finale writes the  `"slashes"` attribute as well as the staff `"number"`, although it does _not_ repeat the underlying content as the spec demands. Its export is definitely better than its import. The software actually only supports one- and two-measure repeats, and thus converts a four-measure repeat into two consecutive two-measure repeats on import (not good). Strangely, it also gets confused when importing its own exports, getting the end measure wrong. 
- MuseScore up to now has had no support for importing or exporting (and, of course, only one-measure repeats were supported in the software). I have implemented export for one-measure repeats as well as the new subtypes, including the `"slashes"` attribute (although I have yet to see a legitimate case where it is meaningful), the `"number"` attribute in the `<measure-style>` tag, _and_ repeating all the content. This export is read perfectly by Finale (Sibelius has its own problems), and of course by the new import I have implemented. This import ignores the `"slashes"` attribute if present, as there is for our purposes an identity between the slashes in the symbol and the number of measures it indicates. But it reads both Sibelius and Finale export correctly (or, in the case of Sibelius, it reads it at least as well as Finale does).

Everything is thoroughly commented and commits have been squashed. I would welcome review from MusicXML specialist @lvinken.

IsaacWeiss commented:
Btw, for easier review of the export commit, enable "Hide whitespace changes." Most of `writeMeasureTracks()` was just indented one additional level.

lvinken commented:
I am perfectly willing to review, but I am severely time-limited at the moment. Most likely I will be unable to spend any time on this for about six weeks :-(.

lvinken commented:
Due to an unexpected change of plans, I may have time to look into this on short notice after all. will keep you posted.

lvinken commented:
Initial results (after clean build on MacOS):
tst_mxml_io fails on:
- testBreaksMMRest.xml
- testMultiMeasureRest1.xml
- testMultiMeasureRest2.xml
- testMultiMeasureRest3.xml
- testMultiMeasureRest4.xml
- testTboxAboveBelow3.xml
- testTboxWords1.xml
-> this probably indicates bugs are still present in the MusicXML import/export changes.

I also get a consistent crash on exiting MuseScore, but this may very well be unrelated. Stack trace:
Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   org.qt-project.QtQml          	0x0000000119a0e3ad QQmlMetaType::qmlSingletonTypes() + 61
1   org.qt-project.QtQml          	0x00000001199e2231 QQmlEngine::~QQmlEngine() + 81
2   org.qt-project.QtQml          	0x00000001199e23ae QQmlEngine::~QQmlEngine() + 14
3   org.qt-project.QtCore         	0x0000000119f136fb QObjectPrivate::deleteChildren() + 203
4   org.qt-project.QtCore         	0x0000000119f134f7 QObject::~QObject() + 1831
5   org.musescore.MuseScore       	0x000000011024f36e mu::framework::UiEngine::~UiEngine() + 158 (uiengine.cpp:54)
6   org.musescore.MuseScore       	0x000000011024ee75 mu::framework::UiEngine::~UiEngine() + 21 (uiengine.cpp:54)
7   libsystem_c.dylib             	0x00007fff7fceaeed __cxa_finalize_ranges + 351
8   libsystem_c.dylib             	0x00007fff7fceb1fe exit + 55
9   libdyld.dylib                 	0x00007fff7fc3e01c start + 8

Finally, I seem to be unable to find any MusicXML test files. Did you create any ? If you did not, please do so.

IsaacWeiss commented on mu4/domain/importexport/internal/musicxml/exportxml.cpp:
```diff
@@ -363,10 +363,10 @@ class ExportMusicXml
+    void writeMeasureTracks(const Measure* m, const int partIndex, const int strack, const int staves,const bool useDrumset,
+                            FigBassMap& fbMap);
+    void writeMeasure(const Measure* const m, const int idx, const int staffCount, MeasureNumberStateHandler& mnsh,FigBassMap& fbMap,
+                      const MeasurePrintContext& mpc);
```
You've got to be kidding me.

IsaacWeiss commented on mu4/scenes/inspector/models/notation/measurerepeats/measurerepeatsettingsmodel.h:
```diff
@@ -7,7 +7,7 @@ class MeasureRepeatSettingsModel : public AbstractInspectorModel
     Q_OBJECT
 
-    Q_PROPERTY(PropertyItem* numberPosition READ numberPosition CONSTANT)
+    Q_PROPERTY(PropertyItem * numberPosition READ numberPosition CONSTANT)
```
Uh-huh

IsaacWeiss commented:
@lvinken thanks, I've been dealing with a transition to an unfamiliar computer/operating system and only just got everything into place. I'll have a look at those test failures and create tests next week.

lvinken commented:
Hi Isaac, FYI, I also plan to review your code changes to the MusicXML importer/exporter, but haven't found the time to do so yet.

Wrt validating the behaviour, that is somewhat difficult. I find the text on http://usermanuals.musicxml.com/MusicXML/MusicXML.htm#CT-MusicXML-measure-repeat.htm quite vague and hard to understand. Furthermore, I do not have access to Finale. I do have Finale Notepad, which does not allow me to create measure repeats, but at least seems to import them.

IsaacWeiss commented:
@lvinken the stray line causing mtest failures has been fixed. I'll add some new tests during this week.

I think the code changes are now limited enough, and commented in enough detail, that reviewing should be swift and painless when you have the chance (and please do enable "Hide whitespace changes"). But in any case, I think my handling is correct; it's not so much that the MusicXML code I wrote is really in need of review, more that it seemed courteous to ask. So if you don't have time before the close of GSoC in two weeks, don't worry about it. Thanks!

IsaacWeiss commented on mu4/domain/importexport/internal/guitarpro/importgtp-gp6.cpp:
```diff
@@ -1970,12 +1970,32 @@ void GuitarPro6::readBars(QDomNode* barList, Measure* measure, ClefType oldClefI
+                    measure->setMeasureRepeatCount(1, staffIdx);
+                } else if (currentNode.toElement().text().compare("FirstOfDouble")) {
+                    // speculative, untested (due to lacking input files),
+                    // but seems like this should work to import two-measure repeats/"similes"
```
Anyone have a Guitar Pro file with a two-measure simile to test?

lvinken commented:
@IsaacWeiss the courtesy is appreciated. Please note that I consider reviewing your code non-optional, to enable maintenance by different developers. Still to do.

I had a look at your test files (thanks for those), this is what I found:

testMeasureRepeats1.xml:
Last eight measures are shown in MuseScore as two "repeat 4" measures, in Finale NotePad
as four "repeat 2" measures
-> assume Finale NotePad issue, as MusicXML specifies a measure-repeat value of "4"

testMeasureRepeats2.xml:
The last measure in the 2nd staff gets a "repeat 2" in Finale NotePad, but notin MuseScore, where it shows a dotted half A2.
-> suspect the measure-repeat type=stop is incorrectly placed (should be at start of measure)

Both files do not conform to the spec, which states "the actual music being repeated needs to be repeated within the MusicXML file".
-> does MuseScore handle a correct file correctly ? Please add correct test file(s) the can be imported and exported unchanged.

Note that, to prevent dependencies on the layout algorithm, reference files must be exported without layout information (only breaks allowed).

Finally, I always appreciate having a PDF available of what the import result should be, this removes any misunderstanding in how to interpret the MusicMXL files.

IsaacWeiss commented:
@lvinken excellent, thanks!

See [above](#issuecomment-671142715) for a discussion of Finale/Sibelius/MuseScore interoperability. In a nutshell, the other two are both varying degrees of shoddy. Neither one adheres to spec (they both export rests instead of the repeated content), and, indeed, as you found in your testing, Finale gets the end measure wrong when importing *its own export!*

The reason I included those sloppily written test files from the major competitors is to test the "lenient with input and strict with output" rule. In `testMeasureRepeats2.xml`, I believe you're correct that Finale wrote the closing `<measure-repeat type="stop">` in the middle of the measure and isn't equipped to handle that on import. But MuseScore can handle it just fine, and if you try opening `testMeasureRepeats2_ref.xml` (MuseScore' output) in Finale, *that* one will be read correctly.

Wrt the ref files, I had exported them while running MuseScore with `--template-mode`, but apparently that only works for MSCX. Now I've found the Preferences setting, I'll replace them. I will also add a correct file (from MuseScore) for an IO test, and PDFs of all three.

lvinken commented:
@IsaacWeiss I have reviewed MusicXML import/export only (and as I don't know how Github's review function works :-), the result is below in plaintext). Your changes are indeed not very complex and achieve the desired result. Nevertheless, there are still a few improvements possible.

My main concern is with ExportMusicXml::writeMeasureTracks():
- could your changes interfere with tbox handling ?
- could your changes interfere with spanner handling ?
- the manipulation of m versus orig_m strongly suggests this function must be split into a wrapper that manipulates measure pointer and tick and a function that writes measure content (which could be the previous writeMeasureTracks() perhaps with minor changes)
- implementing this split would also make writeMeasureTracks()'s parameter m const * const again
- note: the function's existing implementation was already too long, please don't make it worse
 
The code could further be improved by factoring out some of your changes into smaller functions that document what they do.
- the "measure repeat handling" part in MusicXMLParserPass2::measure() (I know, this function was already too long, please don't make it worse)
- the complex logic in the new if statements in measureStyle() in exportxml.cpp

With sufficient test files in place, refactoring should be a low risk job.

IsaacWeiss commented:
>   *  could your changes interfere with tbox handling ?

If a text frame is attached to a MeasureRepeat measure, it will not be exported; if a text frame is attached to a measure that a MeasureRepeat refers to, the text frame will be repeated along with any other text in the measure.

>   *  could your changes interfere with spanner handling ?

Good catch! It seems to depend on the type of spanner. Voltas do not care if they are in the presence of MeasureRepeats or not; hairpins, likely because they attach to ChordRests rather than measures, do get lost if they are attached to MeasureRepeats. I don't know if this is important to handle.

>   *  the manipulation of m versus orig_m strongly suggests this function must be split into a wrapper that manipulates measure pointer and tick and a function that writes measure content (which could be the previous writeMeasureTracks() perhaps with minor changes)
>   *  implementing this split would also make writeMeasureTracks()'s parameter m const * const again
>   *  note: the function's existing implementation was already too long, please don't make it worse

That makes sense. Will do.

>  *  the "measure repeat handling" part in MusicXMLParserPass2::measure() (I know, this function was already too long, please don't make it worse)
>  *  the complex logic in the new if statements in measureStyle() in exportxml.cpp
I will extract these.

lvinken commented:
With respect tot tboxes and spanners: all corner cases should work correctly, but if you run out of time the implementation of handling these corner cases could probably postponed. I would be disappointed, but I wouldn't call it a showstopper. It would most likely lead to support calls later.

IsaacWeiss commented:
I will see what I can do. I think it may come very naturally at the same time as refactoring `writeMeasureTracks()`—I envision moving some things into an middleman`writeMeasureStaves()` function.

IsaacWeiss commented:
@lvinken I am unsure what the correct behavior is with spanners. I can imagine both of these scenarios:
![1](https://user-images.githubusercontent.com/11616153/90802619-82ea4200-e2e5-11ea-9305-3346bd365c1a.png)
![2](https://user-images.githubusercontent.com/11616153/90802624-854c9c00-e2e5-11ea-9098-a343ad140164.png)
My current code handles 1 but not 2. If I handle 2, though, I don't think it's also possible to handle 1.

lvinken commented:
@IsaacWeiss it seems to me the spanner issue is undefined in the MusicXML specification, but if I were to guess, I would assume that the intent is to export musical content (notes and spanners) as if no measure repeats are present and add the measure repeats on top of that. Given the unclarity in the specification and the behaviour of our competitors, I would not consider this a showstopper.

Jojo-Schmitz commented:
I do understand most of the vtest diffs, all the ones that pertain to mm rest and measure repeats, but the tab one is strange:
![tablature-1-ref](https://user-images.githubusercontent.com/1786669/91019062-b7654300-e5f0-11ea-8ffe-2e466d077ffc.png)
vs
![tablature-1-1](https://user-images.githubusercontent.com/1786669/91019080-bd5b2400-e5f0-11ea-9eb4-cdd2719132a9.png)
diff
![tablature-1-diff](https://user-images.githubusercontent.com/1786669/91019094-c3510500-e5f0-11ea-8cb5-f4132bb86dfd.png)

IsaacWeiss commented:
I would guess that it's a result of my combining new code to conceal rests in MeasureRepeat-related measures with existing code to not draw rests in tab. Probably the rest in the last measure was previously laid out, even if not drawn, and now it is not laid out either.

IsaacWeiss commented:
Ignore that "Some checks were not successful"—all tests (that matter) have passed!

Jojo-Schmitz commented:
simple codestyle fail + victim of master not building properly currently

MarcSabatella commented on libmscore/barline.cpp:
```diff
@@ -156,6 +156,12 @@ static void undoChangeBarLineType(BarLine* bl, BarLineType barType, bool allStav
         Measure* m2 = m->isMMRest() ? m->mmRestFirst() : m;
+        for (int staffIdx = 0; staffIdx < m2->score()->nstaves(); ++staffIdx) {
+            if (m2->isMeasureRepeatGroupWithPrevM(staffIdx)) {
+                MScore::setError(CANNOT_SPLIT_MEASURE_REPEAT);
```
General comment: there are a ton of changes for MuseScore 4 that I have no insight into, so I'm just going to generally assume when I see something like this setError that I've never heard of before, that it's probably fine.

MarcSabatella commented on libmscore/cmd.cpp:
```diff
@@ -2848,6 +2848,124 @@ void Score::cmdAddGrace(NoteType graceType, int duration)
+        }
+        if (m != measures.back()) {
+            m->undoSetNoBreak(true);
+            deleteItem(const_cast<BarLine*>(m->endBarLine())); // also takes care of Pid::REPEAT_END
```
Ive been guilty of using const_cast to get around things too, so I'm not necessarily one to talk, but - of all the things that maybe you shouldn't do with a const object, deleting it seems like it would be up there.  Does it perhaps make sense to reconsider the endBarLine method?

IsaacWeiss replied:
Fair point, will do.

IsaacWeiss replied:
Done

MarcSabatella commented on libmscore/layout.cpp:
```diff
@@ -1531,6 +1531,31 @@ void Score::connectTies(bool silent)
 
+//---------------------------------------------------------
+//   relayoutToggleStyles
+///   force relayout if certain style settings are enabled
```
Somehow this doesn't seem to describe what the function actually appears to do.  This function changes style settings, but it isn't obvious why.

IsaacWeiss replied:
How about:

```
///   for styles that can't properly apply to a score that hasn't been laid out yet,
///   temporarily disable them and then reenable after layout
```

IsaacWeiss replied:
Added comments that should make it much more clear

MarcSabatella commented on libmscore/measure.cpp:
```diff
@@ -49,7 +49,7 @@
 #include "pedal.h"
 #include "pitchspelling.h"
-#include "repeat.h"
+#include "measurerepeat.h"
```
I've commented about this before, but it bears "repeating" here, since GitHUb at first declined to show me the diff on this file because it was so big: this was a big diff that wasn't really a functional change at all.  The bulk of it was just renaming of variables.  If there would be any way to add comments - here on GitHub, not in the source - to save reviewers the trouble of figuring out what has really changed and what has not, it would be a big help.

IsaacWeiss replied:
Luckily, I have all the original commits preserved in another branch, so this should be easy.

IsaacWeiss replied:
Done

MarcSabatella commented on libmscore/navigate.cpp:
```diff
@@ -35,16 +36,13 @@ namespace Ms {
 //---------------------------------------------------------
 
-ChordRest* nextChordRest(ChordRest* cr, bool skipGrace)
+ChordRest* nextChordRest(ChordRest* cr, bool skipGrace, bool skipMeasureRepeatRests)
```
Curious that this was needed here, Ordinary navigation (left/right) seems to still hit the hidden rests, though.  Could you explain more about what should and should not be happening here?

IsaacWeiss replied:
On my setup ordinary navigation with the arrow keys does not hit the hidden rests. Could you possibly have assigned the shortcuts to similar but different actions?

MarcSabatella commented on mscore/scoreview.cpp:
```diff
@@ -2363,7 +2364,7 @@ void ScoreView::cmd(const char* s)
                 Element* el = cv->score()->selection().element();
                 Element* oel = el;
-                if (el && (el->isNote() || el->isRest())) {
+                if (el && (el->isNote() || el->isRest() || el->isMMRest() || el->isMeasureRepeat())) {
```
I might still recommend a utility function "isRestType()" for this.  Comes up just often enough, and I can also imagine someday other types being rolled into this.

IsaacWeiss replied:
All right. What do you think of `isRestFamily()` vs. `isRestType()`?

IsaacWeiss replied:
Done with `isRestFamily()` (because I feel "isRestType" semantically says the same thing as "isRest" given that the function checks the element type).

Jojo-Schmitz commented:
rebase needed...

MarcSabatella commented on libmscore/scoreElement.h:
```diff
@@ -399,7 +399,8 @@ class ScoreElement
 
     virtual bool isElement() const { return false; }   // overridden in element.h
-    bool isChordRest() const { return isRest() || isChord() || isRepeatMeasure() || isMMRest(); }
+    bool isRestFamily() const { return isRest() || isMMRest() || isMeasureRepeat(); }
```
The reason I suggested "Type" here is that I was thinking of BarLineType that we used to check for a few related segments types in the same way.  But, that's a bit different, so maybe not important to mimic its naming.

IsaacWeiss commented:
All done.

IsaacWeiss commented:
Re-rebased
